### DISCUSSION TITLE
multi_thread  Doesn't work

### DISCUSSION THREAD

**Discussion by zbcms** [tokio-rs/axum#1569] [Q&A]
### Version

axum = "0.4"
axum = "0.5"

### Platform

64-bit (Windows)

### Description
Step 1： Firefox  request  -> [ sleep = 2s ]
Step 2：Fast switching Chrome  request  -> [ sleep > 3s ]

add  multi_thread  browser or phone Testing Doesn't work
but apache-jmeter Testing is OK

actix-web use sleep  Testing is OK

```
// #[tokio::main]
 #[tokio::main(flavor = "multi_thread", worker_threads = 10)]
 async fn main() {
 
    // Build our application by creating our router.
    let app = axum::Router::new()
    .route("/",get(root));

    // Run our application as a hyper server on http://localhost:3000.
    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
 }
// basic handler that responds with a static string
async fn root( ) ->String{

    let ten_millis = time::Duration::from_millis(2000);
    println!("\r\nsleep start-----------" );
    thread::sleep(ten_millis);
    println!("sleep stop-----------\r\n" );
   
    let time = SystemTime::now();
 
    format!("Hello, World! {:?}",time)
}

```

**Comment by davidpdrsn** [comment]
Don't use `std::thread::sleep`, since that blocks the async runtime. Use https://docs.rs/tokio/latest/tokio/time/fn.sleep.html instead.

**Reply by matthewgapp** [reply]
I'm not able to reproduce the blocking behavior here. Using the example code above (specifically using `std::thread::sleep()`), if I increase the sleep time to 20 secs, I'm able to simultaneously hit the / endpoint with three different curl commands from three different terminals. The requests start and complete in rapid succession as expected. This seems to indicate that each request is given it's own 'thread', as I'd hope so that heavy workloads in a given endpoint don't block all other endpoints. But I was able to reproduce blocking behavior when the Server future is wrapped in tokio::spawn. See [here](https://github.com/tokio-rs/axum/discussions/1695#discussioncomment-4802462)

Would you mind explaining what's going on @davidpdrsn? Why are requests not blocked by default using .await.unwrap() as you suggested and why are they blocked when wrapping the Server in tokio::spawn()?

**Reply by davidpdrsn** [reply]
I recommend reading https://ryhl.io/blog/async-what-is-blocking/
